### PR TITLE
Patient/doctorsearch

### LINKED ISSUE #44: Patient: Doctor search

### PR DESCRIPTION
**Pull request by VladimirGolosin**
Changes:
- Refactor of repos and services
- Add doctor search functionality
closes #44

### CODE REVIEW

File: HospitalSystem.ConsoleUI/PatientUI.cs [hunk 1]
```diff
@@ -473,25 +476,23 @@ public void UpdateCheckup()
...
-        if (oldDate < _now.AddDays(2))
+        if (oldDate < DateTime.Now.AddDays(2))
         {
             CheckupChangeRequest newRequest = new CheckupChangeRequest(
                 selectedCheckup,
                 CRUDOperation.UPDATE);
             Console.WriteLine("Checkup date is in less than 2 days from now. Change request sent.");
-            _hospital.CheckupChangeRequestRepo.AddOrUpdate(newRequest);
+            _hospital.CheckupChangeRequestService.AddOrUpdate(newRequest);
         }
```
Comment (swopl) [inline HospitalSystem.ConsoleUI/PatientUI.cs:496]: Rename add or update to Upsert if that is what it does

File: HospitalSystem.ConsoleUI/PatientUI.cs [hunk 2]
```diff
@@ -620,15 +621,15 @@ public DateTime SelectTime(DateTime inputDate)
         while (iterationTime.TimeOfDay != Globals.ClosingTime.TimeOfDay)
         {
             Console.WriteLine(highestCheckupIndex + " - " + iterationTime.ToString("HH:mm"));
-            iterationTime = iterationTime.Add(_checkupDuration);
+            iterationTime = iterationTime.Add(Globals._checkupDuration);
             highestCheckupIndex += 1;
         }
 
         System.Console.Write("Please enter a number from the list: ");
         int selectedIndex = ReadInt(0, highestCheckupIndex-1, "Number out of bounds!", "Number not recognized!");
```
Comment (swopl) [inline HospitalSystem.ConsoleUI/PatientUI.cs:630]: TimeSlot class smart maybe?

File: HospitalSystem.ConsoleUI/PatientUI.cs [hunk 3]
```diff
@@ -833,24 +834,31 @@ public void CreateCheckup()
             System.Console.Write(e.Message + " Aborting...");
             return;
         }
-
+
```
Comment (swopl) [inline HospitalSystem.ConsoleUI/PatientUI.cs:837]: Ln. 837 bunch of empty space

File: HospitalSystem.ConsoleUI/PatientUI.cs [hunk 4]
```diff
@@ -877,6 +885,115 @@ public void CreateCheckup()
...
     }
+    /*public int CompareDoctorsByRating(Doctor doctor1, Doctor doctor2)
+    {
+        float rating1 = _hospital.AppointmentService.GetAverageRating(doctor1);
+        if (rating1 == 0){
+            rating1 = -10; 
+        }
+        float rating2 = _hospital.AppointmentService.GetAverageRating(doctor2);
+        return String.Compare(name1, name2);
+    }*/
```
Comment (swopl) [inline HospitalSystem.ConsoleUI/PatientUI.cs:896]: Commented out func. Why here?

File: HospitalSystem.ConsoleUI/PatientUI.cs [hunk 5]
```diff
@@ -877,6 +885,115 @@ public void CreateCheckup()
...
+            filteredDoctors.Sort((doctor1, doctor2)=> String.Compare(doctor1.LastName, doctor2.LastName));
+        }
+        else if (sortOption == "s")
+        {
+            filteredDoctors.Sort((doctor1, doctor2)=> String.Compare(doctor1.Specialty.ToString(), doctor2.Specialty.ToString()));
+        }
+         else if (sortOption == "a")
+        {
+            filteredDoctors.Sort((doctor1, doctor2)=>  _hospital.AppointmentService.GetAverageRating(doctor1).CompareTo(_hospital.AppointmentService.GetAverageRating(doctor2)));
+        }
```
Comment (swopl) [inline HospitalSystem.ConsoleUI/PatientUI.cs:963]: Split off into its own func

File: HospitalSystem.ConsoleUI/PatientUI.cs [hunk 6]
```diff
@@ -877,6 +885,115 @@ public void CreateCheckup()
...
+            case "n":
+                filteredDoctors = _hospital.DoctorRepo.GetManyByName(keyword);
+                break;
+            case "l":
+                filteredDoctors = _hospital.DoctorRepo.GetManyByLastName(keyword);
+                break;
+            case "s":
+                filteredDoctors = _hospital.DoctorRepo.GetManyBySpecialty(keyword);
+                break;
+        }
```
Comment (swopl) [inline HospitalSystem.ConsoleUI/PatientUI.cs:930]: Maybe its own func?

**Review by swopl — CHANGES_REQUESTED** [review]

File: HospitalSystem.Core/People/Doctor/DoctorService.cs [hunk 7]
```diff
@@ -28,6 +28,38 @@ public List<Doctor> GetManyBySpecialty(Specialty specialty)
...
+         public List<Doctor> GetManyByLastName(string keyword)
+        {
+            var doctors = GetAll();
+            var allDoctors =
+                from doctor in doctors.AsQueryable<Doctor>()
+                select doctor;
+
+            List<Doctor> filteredDoctors = allDoctors.ToList().FindAll(doctor => doctor.LastName.Contains(keyword));
+            return filteredDoctors;
+        }
```
Comment (swopl) [inline HospitalSystem.Core/People/Doctor/DoctorService.cs:61]: Bad indent WHY ACCEPT YOUR OWN OVER DEV WHEN NO STUFF

**Review by swopl — CHANGES_REQUESTED** [review]

File: HospitalSystem.Core/Appointments/AppointmentRepository.cs [hunk 8]
```diff
@@ -29,6 +29,11 @@ public IMongoCollection<Checkup> GetCheckups()
         return _dbClient.GetDatabase("hospital").GetCollection<Checkup>("checkups");
     }
 
+     public IMongoCollection<Checkup> GetCheckupsByDoctor()
+    {
+        return _dbClient.GetDatabase("hospital").GetCollection<Checkup>("checkups");
+    }
+
```
[the remark this answers is not in the document]
Comment (VladimirGolosin) [reply HospitalSystem.Core/Appointments/AppointmentRepository.cs:36]: Probably not :)

File: HospitalSystem.Core/Appointments/AppointmentService.cs [hunk 9]
```diff
@@ -368,11 +368,32 @@ public List<Checkup> FindSuitableCheckups(Doctor doctor, DateRange interval, Dat
...
         return GetFirstFewFreeCheckups(doctor,3,user);
     }
+    
+     public List<Checkup> GetCheckupsByDoctor(Doctor doctor)
+        {
+            var checkups = _appointmentRepo.GetCheckups();
+            return
+                (from checkup in checkups.AsQueryable()
+                where doctor.Id == checkup.Doctor.Id
+                select checkup).ToList();        }
```
[the remark this answers is not in the document]
Comment (VladimirGolosin) [reply HospitalSystem.Core/Appointments/AppointmentService.cs:400]: thanks

File: HospitalSystem.Core/Appointments/AppointmentService.cs [hunk 10]
```diff
@@ -368,11 +368,32 @@ public List<Checkup> FindSuitableCheckups(Doctor doctor, DateRange interval, Dat
...
+                sum += checkup.DoctorSurvey.Rating;
+                count+=1;
+            }
+        }
+        if (count == 0)
+        {
+            return 10;
+        }
+        return sum/count;
     }
```
[the remark this answers is not in the document]
Comment (VladimirGolosin) [reply HospitalSystem.Core/Appointments/AppointmentService.cs:420]: ok

File: HospitalSystem.Core/Appointments/CheckupChangeRequestService.cs [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+        public void Delete(CheckupChangeRequest request)
+        {
+            _requestRepo.Delete(request);
+        }
+
+        
+        public void UpdateRequest(int indexId, RequestState state)
+        {   
+           _requestRepo.UpdateRequest(indexId, state);
+        }
```
[the remark this answers is not in the document]
Comment (VladimirGolosin) [reply HospitalSystem.Core/Appointments/CheckupChangeRequestService.cs:50]: fixed

File: HospitalSystem.ConsoleUI/PatientUI.cs (hunk 4 above)
Comment (VladimirGolosin) [reply HospitalSystem.ConsoleUI/PatientUI.cs:896]: its gone

File: HospitalSystem.ConsoleUI/PatientUI.cs (hunk 6 above)
Comment (VladimirGolosin) [reply HospitalSystem.ConsoleUI/PatientUI.cs]: will consider

File: HospitalSystem.Core/People/Doctor/DoctorService.cs (hunk 7 above)
Comment (VladimirGolosin) [reply HospitalSystem.Core/People/Doctor/DoctorService.cs:61]: it has stuff, or maybe its from last pr :D
fixed indent

**Review by swopl — APPROVED** [review]

**Pull request merged**